### ISSUE TITLE
Pyright not following custom property decorator's "instance" dependent behavior

### ISSUE THREAD

**Issue by nebbish** [microsoft/pyright#4480]
**Describe the bug**

When a custom made property decorator has a "get" method that depends on whether it was invoked from the class itself or an instance of that class, the "type" of that property when retrieved should not be a Union of both code paths.

Today, both code paths are considered possible no matter how the property is "gotten" - via instance access, or class access.
Therefore, the code below produces the following error:

    Argument of type "Flag | bool" cannot be assigned to parameter "echo" of type "bool" in function "work" [E]

**To Reproduce**

    class Flag(object):
        def __init__(self, value:bool):
            self.value = value
        def __get__(self, inst, cls):
            if inst is None:
                return self
            return self.value

    def work(echo=False):
        if echo:
            print("working with echo")
        else:
            print("working without echo")

    class App(object):
        flag = Flag(True)
        def main(self):
            print(f"Flag value:  {type(self.flag)}") # prints:  <class 'bool'>
            print(f"Flag object: {type(App.flag)}")  # prints:  <class '__main__.Flag'>
            work(self.flag)                                  # <--------- should NOT produce linter error, but does.

    App().main()

**Expected behavior**

Because the plumbing of invoking property decorators is done by Python itself, I think Pyright could figure out which return type to apply based on how **Python** invokes the decorator.  If the decorator's `__get__` method is invoked manually within the user's code, then the detection proposed here would not apply.

**VS Code extension or command-line**

I'm running Pyright version 1.1.284 inside of a VIM IDE via [coc-pyright](https://github.com/fannheyward/coc-pyright) also version 1.1.284

**Comment by erictraut** [comment]
This isn't something a static type checker can do on its own. To get the behavior you're looking for, you need to specify `@overload` signatures for the `__get__` method. For details about `@overload` refer to [PEP 484](https://peps.python.org/pep-0484/#function-method-overloading) or [the official Python documentation](https://docs.python.org/3/library/typing.html#typing.overload).

t would look something like this:
```python
    @overload
    def __get__(self, inst: None, cls) -> Self: ...
    @overload
    def __get__(self, inst: Self, cls) -> bool: ...
```
